### DISCUSSION TITLE
Is it possible to write zero cost wrappers around doubles, for dimensional annotation purposes?

### DISCUSSION THREAD

**Discussion by soerenwolfers** [dotnet/roslyn#74475] [General] [answered]
My attempts at wrapping doubles in dimensionally typed wrappers are O(5x) slower than raw doubles. 

I have previously asked whether this could be improved at 

https://stackoverflow.com/questions/78696148/wrap-doubles-without-overhead-in-csharp

(without genetics, which I use for the actual dimensional units, but even then it's 4x slower)

Is there anything I could do to speed things up?

If not, would it be a reasonable feature request to ask for improvements here?

My previous attempt (copied from the stackoverflow post where it was recommended I post here), with BenchmarkDotNet code that exhibits a 4x slowdown:

```csharp
using System;
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using NUnit.Framework;

namespace MyBenchmark;

public readonly struct MyDouble
{
    public readonly double Value;

    public MyDouble(double value)
    {
        Value = value;
    }

    public static MyDouble operator +(MyDouble self, MyDouble value)
    {
        return new MyDouble(self.Value + value.Value);
    }

    public static MyDouble operator *(MyDouble self, double value)
    {
        return new MyDouble(self.Value * value);
    }
}

public class DoubleBenchmark
{
    private const int N = 1_000_000;
    private readonly double[] _data;
    private readonly double[] _workspace;
    private readonly MyDouble[] _typedData;
    private readonly MyDouble[] _typedWorkspace;

    public DoubleBenchmark()
    {
        _data= new double[N];
        _workspace = new double[N];
        _typedWorkspace = new MyDouble[N];
        var rand = new Random(0);
        for (var i = 0; i < N; i++)
        {
            _data[i] = rand.NextDouble();
        }
        _typedData = _data.Select(x => new MyDouble(x)).ToArray();
    }

    [Benchmark]
    public MyDouble TypedDoubles()
    {
        var d = new MyDouble(0);
        for (var i = 0; i < N; i++)
        {
            d += _typedData[i] * 3;
            _typedWorkspace[i] = d * 2 + new MyDouble(1);
        }
        var e = new MyDouble(0);
        for (var i = 0; i < N; i++)
        {
            e += _typedWorkspace[i];
        }

        return e;
    }

    [Benchmark]
    public MyDouble ManuallyInlinedTypedDoubles()
    {
        var d = new MyDouble(0);
        for (var i = 0; i < N; i++)
        {
            d = new MyDouble(d.Value + _typedData[i].Value * 3);
            _typedWorkspace[i] = new MyDouble(d.Value * 2 + 1);
        }
        var e = new MyDouble(0);
        for (var i = 0; i < N; i++)
        {
            e = new MyDouble(e.Value + _typedWorkspace[i].Value);
        }

        return e;
    }

    [Benchmark]
    public double Doubles()
    {
        var d = 0d;
        for (var i = 0; i < N; i++)
        {
            d += _data[i] * 3;
            _workspace[i] = d * 2 + 1;
        }
        var e = 0d;
        for (var i = 0; i < N; i++)
        {
            e += _workspace[i];
        }
        
        return e;
    }
    
    [Test, Explicit]
    public static void RunBenchmarks()
    {
        BenchmarkRunner.Run<DoubleBenchmark>();
    }
    
    [Test, Explicit]
    public static void EqualResults()
    {
        var a = new DoubleBenchmark();
        Console.WriteLine((a.TypedDoubles().Value, a.Doubles(), a.ManuallyInlinedTypedDoubles().Value));
        if (Math.Abs(a.TypedDoubles().Value - a.Doubles()) > 0 ||
            Math.Abs(a.TypedDoubles().Value - a.ManuallyInlinedTypedDoubles().Value) > 1e-16)
        {
            throw new Exception("Different results");
        }
    }
}
```

**Comment by CyrusNajmabadi** [comment]
This question lacks necessary information. What approach are you currently taking?  Where is it slow?  What part of that slowness is due to Roslyn?

**Reply by soerenwolfers** [reply]
Hi, I've copied the benchmark code I was running into this post (this is copied from my stackoverflow question). I don't know where it is slow and to what extent this is due to Roslyn.

**Reply by CyrusNajmabadi** [reply]
Closing out as out of scope.  If you determine this is due to anything roslyn is doing, please feel free to open an issue with the relevant information.  

Tagging @tannergooding in case he has some ideas though.  My guess is that this isn't related to anything Roslyn does (we should be emitting idiomatic IL for all of that).  So it's possibly due to how the runtime analyzes and optimize simple struct wrappers.

**Reply by tannergooding** [reply, marked as answer]
Zero-cost wrappers in general are a complex topic and there is no definitive answer because it is often dependent on the ABI (application binary interface) that exists for the underlying C runtime.

In general, `struct S { T x; }` and `T` are not equivalent at the ABI level and while they often do get passed around the same, there are many cases that they may not and that can lead to deficiencies.

It used to be that structs with single-float fields could not be "promoted" (an optimization) due to such ABI differences, but I resolved that in the .NET 8 timeframe: https://github.com/dotnet/runtime/pull/84627.  -- This impacted all single-field floating-point wrapper structs, regardless of whether inlining happened or not.

Even with that fix, however, there is still an ABI difference due to it being passed in an integer register rather than a floating-point register and thus causing data to need to be moved from the floating-point and into the general-purpose registers and then back again, across the call boundary (which is only an issue if inlining can't occur).

There's some other general pessimizations that can happen as well if you aren't careful (but which are again mostly impactful if inlining can't occur). For example, instance methods require a hidden `this` parameter which effectively requires passing the data byref, if the method isn't inlined this causes the value to be spilled to the stack and marked as `address taken` (this is an issue in C/C++ as well). So you really need inlining to occur and for the JIT to undo the `address taken` consideration, elide the copy, etc to get good codegen (it does normally do all this).

-------------------

In this case, .NET 8 and .NET 9 both give numbers similar to the following, so my expectation is that you're running on .NET 6, is that correct?
```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.26100.1150)
AMD Ryzen 9 7950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 9.0.100-preview.6.24328.19
  [Host]     : .NET 9.0.0 (9.0.24.32707), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  DefaultJob : .NET 9.0.0 (9.0.24.32707), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI


| Method                      | Mean     | Error     | StdDev    |
|---------------------------- |---------:|----------:|----------:|
| TypedDoubles                | 1.260 ms | 0.0251 ms | 0.0269 ms |
| ManuallyInlinedTypedDoubles | 1.256 ms | 0.0140 ms | 0.0131 ms |
| Doubles                     | 1.233 ms | 0.0042 ms | 0.0035 ms |
```

.NET 6 gives the following, which is expected due to the lack of struct promotion:
```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.26100.1150)
AMD Ryzen 9 7950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 9.0.100-preview.6.24328.19
  [Host]     : .NET 6.0.30 (6.0.3024.21525), X64 RyuJIT AVX2
  DefaultJob : .NET 6.0.30 (6.0.3024.21525), X64 RyuJIT AVX2


| Method                      | Mean     | Error     | StdDev    |
|---------------------------- |---------:|----------:|----------:|
| TypedDoubles                | 7.374 ms | 0.1466 ms | 0.2529 ms |
| ManuallyInlinedTypedDoubles | 4.622 ms | 0.0220 ms | 0.0206 ms |
| Doubles                     | 1.259 ms | 0.0061 ms | 0.0057 ms |
```

**Reply by soerenwolfers** [reply]
@tannergooding Thanks for your brilliant and comprehensive answer. I am on .NET6 indeed and am excited to hear about the significant performance gains to come for free when we eventually move onto .NET8
